pimd/pim_cmd_common.c

```diff
@@ -653,3 +673,1302 @@ int pim_process_no_rp_plist_cmd(struct vty *vty, const char *rp_str,
+			pim_ifchannel_ifjoin_name(ch->ifjoin_state, ch->flags),
+			uptime, expire, prune);
+	}
+}
+
+void pim_show_join(struct pim_instance *pim, struct vty *vty, pim_sgaddr *sg,
+		   json_object *json)
+{
+	struct pim_interface *pim_ifp;
+	struct pim_ifchannel *ch;
+	struct interface *ifp;
+	time_t now;
+
+	now = pim_time_monotonic_sec();
+
+	if (!json)
+		vty_out(vty,
+			"Interface        Address         Source          Group           State      Uptime   Expire Prune\n");
+
+	FOR_ALL_INTERFACES (pim->vrf, ifp) {
+		pim_ifp = ifp->info;
+		if (!pim_ifp)
+			continue;
+
+		RB_FOREACH (ch, pim_ifchannel_rb, &pim_ifp->ifchannel_rb) {
+			if (!pim_sgaddr_match(ch->sg, *sg))
+				continue;
+
+			pim_show_join_helper(vty, pim_ifp, ch, json, now);
+		} /* scan interface channels */
+	}
+}
+
+static void pim_show_jp_agg_helper(struct vty *vty, struct interface *ifp,
+				   struct pim_neighbor *neigh,
+				   struct pim_upstream *up, int is_join)
+{
+	vty_out(vty, "%-16s %-15pPAs %-15pPAs %-15pPAs %5s\n", ifp->name,
+		&neigh->source_addr, &up->sg.src, &up->sg.grp,
+		is_join ? "J" : "P");
+}
+
+void pim_show_jp_agg_list(struct pim_instance *pim, struct vty *vty)
+{
+	struct interface *ifp;
+	struct pim_interface *pim_ifp;
+	struct listnode *n_node;
+	struct pim_neighbor *neigh;
+	struct listnode *jag_node;
+	struct pim_jp_agg_group *jag;
+	struct listnode *js_node;
+	struct pim_jp_sources *js;
+
+	vty_out(vty,
+		"Interface        RPF Nbr         Source          Group           State\n");
+
+	FOR_ALL_INTERFACES (pim->vrf, ifp) {
+		pim_ifp = ifp->info;
+		if (!pim_ifp)
+			continue;
+
+		for (ALL_LIST_ELEMENTS_RO(pim_ifp->pim_neighbor_list, n_node,
+					  neigh)) {
+			for (ALL_LIST_ELEMENTS_RO(neigh->upstream_jp_agg,
+						  jag_node, jag)) {
+				for (ALL_LIST_ELEMENTS_RO(jag->sources, js_node,
+							  js)) {
+					pim_show_jp_agg_helper(vty, ifp, neigh,
+							       js->up,
+							       js->is_join);
+				}
+			}
+		}
+	}
+}
+
+static void pim_show_membership_helper(struct vty *vty,
+				       struct pim_interface *pim_ifp,
+				       struct pim_ifchannel *ch,
+				       struct json_object *json)
+{
+	char ch_grp_str[PIM_ADDRSTRLEN];
+	json_object *json_iface = NULL;
+	json_object *json_row = NULL;
+
+	json_object_object_get_ex(json, ch->interface->name, &json_iface);
+	if (!json_iface) {
+		json_iface = json_object_new_object();
+		json_object_pim_ifp_add(json_iface, ch->interface);
+		json_object_object_add(json, ch->interface->name, json_iface);
+	}
+
+	snprintfrr(ch_grp_str, sizeof(ch_grp_str), "%pPAs", &ch->sg.grp);
+
+	json_row = json_object_new_object();
+	json_object_string_addf(json_row, "source", "%pPAs", &ch->sg.src);
+	json_object_string_addf(json_row, "group", "%pPAs", &ch->sg.grp);
+	json_object_string_add(json_row, "localMembership",
+			       ch->local_ifmembership == PIM_IFMEMBERSHIP_NOINFO
+					   ? "NOINFO"
+					   : "INCLUDE");
+	json_object_object_add(json_iface, ch_grp_str, json_row);
```

Comment:
`json_object_object_addf` + remove extra step to format `ch_grp_str`

Reply:
Resolved